### PR TITLE
Display relative date time in conversation list

### PR DESCRIPTION
**Pull request by danrabbit**
This duplicates the functionality from https://github.com/elementary/mail/blob/master/src/client/util/util-date.vala

Show messages received:
- [x] Less than a minute ago as "Now"
- [x] Less than an hour ago as "5m ago"
- [x] Today, Less than 12 hours ago as "1h ago"
- [x] All other times Today as "10:04 AM"
    - [x] With localized 12/24 hours
- [x] Yesterday as "Yesterday"
- [x] Less than a week ago as "Sat"
- [x] This year as "Jun 15"
- [x] All other times as "nn/nn/nn", localized

### CODE REVIEW

File: src/ConversationList/ConversationListItem.vala [hunk 1]
```diff
@@ -114,4 +114,45 @@ public class Mail.ConversationListItem : Gtk.ListBoxRow {
...
+            }
+        }
+        if (is_same_day (date_time.add_days (1), now)) {
+            return _("Yesterday");
+        }
+        if (diff < 6 * TimeSpan.DAY) {
+            return date_time.format ("%a");
+        }
+        if (date_time.get_year () == now.get_year ()) {
+            return date_time.format ("%b %-e");
```
Comment (tintou) [inline src/ConversationList/ConversationListItem.vala]: this needs to be localized

File: src/ConversationList/ConversationListItem.vala [hunk 2]
```diff
@@ -114,4 +114,45 @@ public class Mail.ConversationListItem : Gtk.ListBoxRow {
...
+            if (diff < 12 * TimeSpan.HOUR) {
+                int rounded = (int) Math.round ((double) diff / TimeSpan.HOUR);
+                return ngettext ("%dh ago", "%dh ago", (ulong) rounded).printf (rounded);
+            }
+            var settings = new Settings ("org.gnome.desktop.interface");
+            if (settings.get_enum ("clock-format") == 0) {
+                return date_time.format ("%H:%M");
+            } else {
+                return date_time.format ("%l:%M %p");
+            }
```
Comment (tintou) [inline src/ConversationList/ConversationListItem.vala]: there is a Granite util that takes care of all of this for you

File: src/ConversationList/ConversationListItem.vala [hunk 3]
```diff
@@ -114,4 +114,45 @@ public class Mail.ConversationListItem : Gtk.ListBoxRow {
...
+    private string get_relative_datetime (DateTime date_time) {
+        var now = new DateTime.now_local ();
+        var diff = now.difference (date_time);
+
+        if (is_same_day (date_time, now)) {
+            if (diff < TimeSpan.MINUTE) {
+                return _("Now");
+            }
+            if (diff < TimeSpan.HOUR) {
+                return ngettext ("%dm ago", "%dm ago", (ulong) (diff / TimeSpan.MINUTE)).printf ((int) (diff / TimeSpan.MINUTE));
```
Comment (tintou) [inline src/ConversationList/ConversationListItem.vala]: Don't divide twice, store the value and reuse it instead

File: src/ConversationList/ConversationListItem.vala [hunk 4]
```diff
@@ -114,4 +114,45 @@ public class Mail.ConversationListItem : Gtk.ListBoxRow {
...
     }
+
+    private string get_relative_datetime (DateTime date_time) {
+        var now = new DateTime.now_local ();
+        var diff = now.difference (date_time);
+
+        if (is_same_day (date_time, now)) {
+            if (diff < TimeSpan.MINUTE) {
+                return _("Now");
+            }
```
Comment (tintou) [inline src/ConversationList/ConversationListItem.vala]: Use } else if
even if the return doesn't make a difference, it makes if clearer when you read it

File: src/ConversationList/ConversationListItem.vala [hunk 5]
```diff
@@ -114,4 +114,45 @@ public class Mail.ConversationListItem : Gtk.ListBoxRow {
...
+            if (diff < TimeSpan.HOUR) {
+                return ngettext ("%dm ago", "%dm ago", (ulong) (diff / TimeSpan.MINUTE)).printf ((int) (diff / TimeSpan.MINUTE));
+            }
+            if (diff < 12 * TimeSpan.HOUR) {
+                int rounded = (int) Math.round ((double) diff / TimeSpan.HOUR);
```
Comment (tintou) [inline src/ConversationList/ConversationListItem.vala]: another way to calculate it is:
(diff - (diff % TimeSpan.HOUR)) / TimeSpan.HOUR
diff % TimeSpan.HOUR returns diff - max (X * TimeSpan.HOUR) and thus the division doesn't need to be truncated afterward

File: src/ConversationList/ConversationListItem.vala (hunk 2 above)
Comment (tintou) [inline src/ConversationList/ConversationListItem.vala]: https://valadoc.org/granite/Granite.DateTime.html

File: src/ConversationList/ConversationListItem.vala (hunk 3 above)
Comment (tintou) [inline src/ConversationList/ConversationListItem.vala]: also you want to do
`diff - diff % TimeSpan.MINUTE` before to remove all the seconds, and milliseconds parts

File: src/ConversationList/ConversationListItem.vala (hunk 5 above)
Comment (eyelash) [inline src/ConversationList/ConversationListItem.vala]: @tintou but `Math.round()` does not simply truncate. It rounds to the nearest integer, so 1.9 hours will be rounded to 2.

File: src/ConversationList/ConversationListItem.vala (hunk 3 above)
Comment (eyelash) [inline src/ConversationList/ConversationListItem.vala]: > also you want to do
> `diff - diff % TimeSpan.MINUTE` before to remove all the seconds, and milliseconds parts

No need for that. TimeSpan is an integral type and integer division will be automatically truncated.

Comment (tintou) [inline src/ConversationList/ConversationListItem.vala]: Right, the first comment remains valid

File: src/ConversationList/ConversationListItem.vala (hunk 5 above)
Comment (tintou) [inline src/ConversationList/ConversationListItem.vala]: Yeah, which seems wrong to me, we don't want to round but to truncate here

Comment (danrabbit) [inline src/ConversationList/ConversationListItem.vala]: I disagree I think we do want to round. If the actually number of hours is 1.9 and we display 1h we're much more incorrect than if we say 2h

**Pull request merged**